apps/emqx_durable_storage/src/emqx_ds_storage_skipstream_lts.erl

```diff
@@ -0,0 +1,739 @@
+                                    {ok, NextTS, Key, Msg0};
+                                false ->
+                                    %% Hash collision. Advance to the
+                                    %% next timestamp:
+                                    inc_counter(?DS_SKIPSTREAM_LTS_HASH_COLLISION),
+                                    {seek, NextTS + 1}
+                            end;
+                        _ ->
+                            %% This is index stream. Keep going:
+                            next_step(S, StaticIdx, CompressedTF, Iterators, NextTS, {seek, NextTS})
+                    end;
+                NextTS when NextTS > ExpectedTS, N > 0 ->
+                    inc_counter(?DS_SKIPSTREAM_LTS_MISS),
+                    %% Next index level is not what we expect.
+                    {seek, NextTS}
+            end
+    end.
+
+free_iterators(Its) ->
+    lists:foreach(
+        fun(#l{handle = IH}) ->
+            ok = rocksdb:iterator_close(IH)
+        end,
+        Its
+    ).
+
+%%%%%%%% Indexes %%%%%%%%%%
+
+mk_index(HashBytes, Static, Timestamp, Varying) ->
+    mk_index(HashBytes, Static, Timestamp, 1, Varying, []).
+
+mk_index(HashBytes, Static, Timestamp, N, [TopicLevel | Varying], Acc) ->
+    Op = {mk_key(Static, N, hash(HashBytes, TopicLevel), Timestamp), <<>>},
+    mk_index(HashBytes, Static, Timestamp, N + 1, Varying, [Op | Acc]);
+mk_index(_HashBytes, _Static, _Timestamp, _N, [], Acc) ->
+    Acc.
+
+%%%%%%%% Keys %%%%%%%%%%
+
+get_key_range(StaticIdx, WildcardIdx, Hash) ->
+    [
+        {iterate_lower_bound, mk_key(StaticIdx, WildcardIdx, Hash, 0)},
+        {iterate_upper_bound, mk_key(StaticIdx, WildcardIdx, Hash, ?max_ts)}
+    ].
+
+-spec match_ds_key(emqx_ds_lts:static_key(), binary()) -> ts() | false.
+match_ds_key(StaticIdx, Key) ->
+    match_key(StaticIdx, 0, <<>>, Key).
+
+-spec match_key(emqx_ds_lts:static_key(), wildcard_idx(), wildcard_hash(), binary()) ->
+    ts() | false.
+match_key(StaticIdx, 0, <<>>, Key) ->
+    TSz = size(StaticIdx),
+    case Key of
+        <<StaticIdx:TSz/binary, 0:?wcb, Timestamp:?tsb>> ->
+            Timestamp;
+        _ ->
+            false
+    end;
+match_key(StaticIdx, Idx, Hash, Key) when Idx > 0 ->
+    Tsz = size(StaticIdx),
+    Hsz = size(Hash),
+    case Key of
+        <<StaticIdx:Tsz/binary, Idx:?wcb, Hash:Hsz/binary, Timestamp:?tsb>> ->
+            Timestamp;
+        _ ->
+            false
+    end.
+
+-spec mk_key(emqx_ds_lts:static_key(), wildcard_idx(), wildcard_hash(), ts()) -> binary().
+mk_key(StaticIdx, 0, <<>>, Timestamp) ->
+    %% Data stream is identified by wildcard level = 0
+    <<StaticIdx/binary, 0:?wcb, Timestamp:?tsb>>;
+mk_key(StaticIdx, N, Hash, Timestamp) when N > 0 ->
+    %% Index stream:
+    <<StaticIdx/binary, N:?wcb, Hash/binary, Timestamp:?tsb>>.
+
+hash(HashBytes, '') ->
+    hash(HashBytes, <<>>);
+hash(HashBytes, TopicLevel) ->
+    {Hash, _} = split_binary(erlang:md5(TopicLevel), HashBytes),
```

Comment:
Tiny nit. While `<<Hash:HashBytes/bytes, _/binary>>` looks uglier, it doesn't require allocating a tuple on the heap.